### ISSUE TITLE
Make generated classes extend common super-type

### ISSUE THREAD

**Issue by nutgaard** [ExpediaGroup/graphql-kotlin#800]
**Is your feature request related to a problem? Please describe.**
I want to create a function where some common transforms/error-handling/logging can live. 
As of now each class exists independently in the class hierarchy, and thus it is non-trivial to create abstractions with them.

Given the example classes here;
```kotlin
class FetchName(private val graphQLClient: GraphQLClient<*>) { ... } // Generated for fetchName.graphql
class FetchNameBulk(private val graphQLClient: GraphQLClient<*>) { ... } // Generated for fetchNameBulk.graphql
```

In order to pack some custom-logic in here (mainly custom error-handling and logging) I want to do something like this;
```kotlin
fun exec(query: FetchName, variables: FetchName.Variables): GraphlQLResponse<FetchName.Result>? {
  val uuid = UUID.randomUUID();
  return try {
    val transformedVariables = transform(variables)
    servicecallLog.info(uuid, "FetchNameRequest", transformedVariables)
    val response = query.execute(transformedVariables)
    if (response.errors.isNullOrEmpty()) {
      servicecallLog.info(uuid, "FetchNameResponse", response)
      response
    } else {
      servicecallLog.error(uuid, "FetchNameError", response.errors)
      null
    }
  } catch (exception: Exception) {
    servicecallLog.error(uuid, "FetchNameError", e)
    null
  }
}
```

Now the code can be duplicated and changed to fit for `FetchNameBulk` as well, but I would prefer a more generic approach. 

**Describe the solution you'd like**
Experimented a bit with the code-generation, and how it could be structured. That way I found that introducing a common super-interface could solve my issues pretty neatly without (as far as I could tell) introducing any adverse effects.

What I propose is something along the lines of;
```kotlin
class FetchName(private val graphQLClient: GraphQLClient<*>) : GraphQLRequestDescription<FetchName.Variables, FetchName.Result> { ... }
```

That way the `exec` function above could be generalize 
```kotlin
fun <VARS, RESULT> exec(query: GraphQLRequestDescription<VARS, RESULT>, variables: VARS): GraphQLResponse<RESULT>
```

Personally I would prefer `GraphQLRequest` instead of `GraphQLRequestDescription`, but since it's already used in the types-package I used `GraphQLRequestDescription` when testing.

**Describe alternatives you've considered**
The "problem" can be though of as three separate issues; variable-transforms, logging and custom-errorhandling.

Errorhandling is trivial to solve as the return-type already has a common return type `GraphQLResponse<T>`.  

For logging I looked into using ktor-client-logging, but from what I could understand you are pretty limited in what you can do with it. Another possibility is to create a custom `HttpClientFeature`, but again it makes it hard to log just the graphql-variables for instance.

Custom variable-transforms could be hoisted to the callsite of the `exec` function. But I would then have to rely on every developer to call the transform-function before passing the variables.

**Addition context**
I've played around with some code locally, and believe I may have working proof-of-concept (absent the tests).
But wanted to create a issue to discuss the feature before diving even deeper into the issue or create a PR.
The changes can be seen here; https://github.com/nutgaard/graphql-kotlin/commit/4d8887878f372d521877801cc09549b952c125de

**Comment by smyrick** [comment]
@nutgaard We specifically chose Ktor for its customizations options. So I think what you are looking for could be accomplished with a custom [Ktor Client Feature](https://ktor.io/clients/http-client/features.html). While this may be new code to write, we knew this when creating the client that it was not going to be fully featured, but the community would create the features they need.

If you do have a custom feature that exists today that just depends on the `graphql-kotlin-types` packages would be happy to link to that in our docs as a community extension

**Comment by nutgaard** [comment]
I completely understand wanting to keep the library as simple as possible, and pushing customization in general to the ktor-client. 
And while writing custom client-features may be a suitable approach for alot of features, I'm not sure everything can be easily solved with ktor-client-features. I tried looking into before diving into sourcecode of the library, but actually found making changes to the library through kotlinpoet easier, so, make of that what you want. :P 

I did however look into to once more, and while the logging seems to be solvable if you disregard wanting to just logging the variables. 
It would seem that doing a transformation of the variables would involve some really nasty hacks as far as I could tell. :/ 

The way I see it this would add a nice extension-point to the library, allowing devs to develop composable pieces of code in an easy way. 

What's not to like? ;)

**Comment by dariuszkuc** [comment]
I'm thinking we probably could refactor existing code such that `GraphQLClient` becomes an interface and then have specific implementations (e.g. existing `KtorGraphQLClient` and potentially something like `SpringGraphQLClient`). You could then create your own implementation of the clients that override the defaults.

If we decide to do this sort of refactor it will take place in `4.0.0` release.

**Comment by nutgaard** [comment]
@dariuszkuc Such a refactoring is probably something people will appreciate. 

Maybe I'm misunderstanding or simply missing something, but I'm not quite sure how it would solve the problem I'm facing. `FetchName` and `FetchNameBulk` would still be independency classes in the class-hierarchi. Thus making a new GraphQLClient wouldn't easily be able to work with both of them. :/ 

What I'm I missing here? :/

**Comment by dariuszkuc** [comment]
with the proposed refactor you would move that custom logic to your client, e.g.

```kotlin
class CustomGraphQLClient : KtorGraphQLClient {

  override suspend fun <T> execute(query: String, operationName: String? = null, variables: Any? = null, resultType: Class<T>, requestBuilder: HttpRequestBuilder.() -> Unit = {}): GraphQLResponse<T> {
      // your custom logic
      super.execute(query, operationName, variables, resultType, requestBuilder)
      // your finalization logic
  }
}
```

**Comment by nutgaard** [comment]
Of course. that would work. 🤦 

My love for generics kind of still wishes for the super-interface approach, but I've no problems understanding why you would want this approach instead. It does solve all of my problems, so I'll be happy either way. 

Anyways, we currently have a working solution in our codebase. But the code is a bit messy (hence this issue), so I'll be eagerly waiting for the `4.0.0` release. ;)

**Comment by dariuszkuc** [comment]
I looked into updating `GraphQLClient` into a generic interface and while its possible to do it we would loose some of the existing Ktor customizability, i.e. current execute method signature accepts `HttpRequestBuilder.()` lambda which allows us to apply per request customization. If we move to generic interface then we would need to provide some generic mechanism to duplicate the same (or part of) functionality (e.g.. maybe explicitly accept list of headers) that could then be applied in the specific implementations. I'm going to icebox generic client interface until we figure out a nice way to generically apply those customizations.

In the meantime I just opened up the client (#811) so you could still extend it and create your custom client.

**Comment by dariuszkuc** [comment]
Change was released in `3.5.0`
